[CORE-13480] kafka/client: improve topic caching behavior

IoannisRP opened the pull request:
Implements: [CORE-13480](https://redpandadata.atlassian.net/browse/CORE-13480)

Change how updates to topic cache are being applied. Instead of the new data overwriting the old ones, they are merged. This happens according to:
- A partition is not updated if the old state has higher leader epoch
- A topic is erased from the cache if it isn't part of N consecutive updates (arbitrarily set to 5)
- If a topic changes ID, the old data are discarded in favor or the new
- If a requested topic returns an error, it is explicitly removed from the topic cache.

Also, the list of topics to request metadata for is no longer part of cluster's api, but it is provided through the configuration.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->


[CORE-13480]: https://redpandadata.atlassian.net/browse/CORE-13480?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
kafka/client: Improve topic_cache update logic

Change how updates to topic cache are being applied. Instead of the new
data overwritting the old ones, they are merged. This happens according
to:
- A partition is not updated if the old state has higher leader epoch
- A topic is erased from the cache if it isn't part of N consecutive
  updates (arbitrarily set to 5)
- If a topic changes ID, the old data are discarded in favor or the new

IoannisRP committed:
kafka/client: Add validation in topic_cache update

If the response for a topic is not successful, it should be ignored

michael-redpanda commented on src/v/kafka/client/topic_cache.h:
```diff
@@ -67,9 +68,21 @@ class topic_cache {
+
     /// \brief Cache of topic information.
     topics_t _topics;
+    topic_counters_t _missing_topic_counters;
```
question: why not have this as part of `topic_cache::topic_data`?

michael-redpanda replied:
I think this would be easier to manage than having two copies of the topic name and having to manage two separate maps

IoannisRP replied:
> question: why not have this as part of topic_cache::topic_data?

cause i didn't think of that :smile:

michael-redpanda commented on src/v/kafka/client/topic_cache.cc:
```diff
@@ -49,7 +51,83 @@ void topic_cache::apply(
+        auto it = _topics.find(topic);
+        if (it == _topics.end()) {
+            // topic doesn't have cached data. Accept new value
+            merged.emplace(topic, std::move(updated_topic));
```
nit you std::move(topic) below but not here

IoannisRP replied:
i missed the other one. My aim was to not move any key, as messing with the keys of a map (theoretically) messes up its assumptions.

For std map(s), keys are const. As i found out, we don't do the same on our maps.

michael-redpanda commented on src/v/kafka/client/topic_cache.h:
```diff
@@ -67,9 +68,21 @@ class topic_cache {
 
     const topics_t& cache() const noexcept;
 
+    static constexpr size_t missing_topic_limit = 5;
```
I think we do a lot of metadata refreshes and I'd be worried about this limit being too small and accidently clearing out topics when unnecessary.  One thought I had was to use a 'last seen' timestamp and if we haven't seen it in >30 seconds or >30 minutes we'd clear it.

IoannisRP replied:
went for a default value of 60s

michael-redpanda commented on src/v/kafka/client/configuration.h:
```diff
@@ -101,6 +102,10 @@ struct connection_configuration {
+    // std::nullopt - subscribed to all topics on the cluster
+    // otherwise only subscribed to the list
+    std::optional<chunked_vector<model::topic>> subscribed_topics
+      = std::nullopt;
```
do we need this?  I think this is a nice to have but not necessary for the changes required for the topic cache improvements.

michael-redpanda commented on src/v/cluster_link/security_migrator.cc:
```diff
@@ -288,7 +288,7 @@ ss::future<task::state_transition> security_migrator::run_impl() {
 
     try {
-        co_await cluster.request_metadata_update(std::nullopt);
+        co_await cluster.request_metadata_update();
```
This was somewhat deliberate as the security migrator doesn't really care about any topic information.

IoannisRP replied:
do you mean it should not get info for *any* topic? In that case, it should request an empty list, not a nullopt :thinking:

michael-redpanda commented on src/v/kafka/client/cluster.cc:
```diff
@@ -423,6 +424,20 @@ ss::future<> cluster::apply_metadata(metadata_update reply) {
+                    _topic_cache.erase(topic);
+                }
+            }
+        }
```
I'm not sure we should do that.  Why not just let the topic timeout?

michael-redpanda replied:
I think we should err on the side of caution and not remove topics even if they have an error - could be transient (e.g. broker not available or somethign)

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5a1fc32f-7488-43ac-86ce-c667bdd5f8cf-->
<!--TEST RESULTS-->
<!--cad8f7e0-6e5e-4a97-b7e2-688be4809e24-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75562">build#75562</a></summary>


| test_class                          | test_method                                  | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:------------------------------------|:---------------------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalTest          | test_tx_control_batch_removal                | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75562#019a4fe9-c82a-4de3-9424-58babf8b6879 | FLAKY         | 14/21    | upstream reliability is '96.26038781163435'. current run reliability is '66.66666666666666'. drift is 29.59372 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal   |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75562#019a4fe9-c828-4f59-afb6-446df9149061 | FLAKY         | 20/21    | upstream reliability is '94.71744471744472'. current run reliability is '95.23809523809523'. drift is -0.52065 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75562#019a4fe9-c82b-4870-9995-f88b60ad10e0 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
| WriteCachingFailureInjectionE2ETest | test_crash_all                               | {"use_transactions": false}                        | integration | https://buildkite.com/redpanda/redpanda/builds/75562#019a501a-1a7b-455c-869e-5678bd67efc4 | FLAKY         | 17/21    | upstream reliability is '88.34080717488789'. current run reliability is '80.95238095238095'. drift is 7.38843 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75667">build#75667</a></summary>


| test_class                          | test_method                                             | test_arguments                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                                   |
|:------------------------------------|:--------------------------------------------------------|:------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTest                   | compaction_stress_test                                  | {"cleanup_policy": "compact,delete", "key_set_cardinality": 100, "storage_compaction_key_map_memory_kb": 3} | integration | https://buildkite.com/redpanda/redpanda/builds/75667#019a5510-1cf8-4aa2-b945-491b0da51f43 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTest&test_method=compaction_stress_test                                         |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer            | null                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75667#019a5510-1cf5-4f61-a0b4-520cd0f6d783 | FLAKY         | 19/21    | upstream reliability is '94.56398640996603'. current run reliability is '90.47619047619048'. drift is 4.0878 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer                       |
| NodesDecommissioningTest            | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": false}                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75667#019a5510-1cf8-4aa2-b945-491b0da51f43 | FLAKY         | 18/21    | upstream reliability is '98.0806142034549'. current run reliability is '85.71428571428571'. drift is 12.36633 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                          | {"use_transactions": false}                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/75667#019a5510-1cfa-41d8-aaec-656a5b8c0568 | FLAKY         | 16/21    | upstream reliability is '89.12058023572075'. current run reliability is '76.19047619047619'. drift is 12.9301 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                               |
</details>
<!--/body-->

IoannisRP committed:
kafka/client: Add topic timeout in topic cache

If a topic hasn't received an update for longer than the specified
timeout, it is removed from the cache.

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/954dfae00247b09cece009778f3e0ef8a863580a..c603f6b0f95c71fa0114aa948f9100e27644b2cc):
- rework logic to a timer instead of a counter for when topics are removed
- removed subscription logic

changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c603f6b0f95c71fa0114aa948f9100e27644b2cc..e19867a7c8d734508fbbdae8d8ca2e58501a55b6):
- rebase to dev

IoannisRP commented on src/v/kafka/client/topic_cache.h:
```diff
@@ -67,9 +72,24 @@ class topic_cache {
+
+    /// \brief If a topic hasn't been updated in this much time, it will be
+    /// considered stale and removed
+    std::chrono::seconds _topic_timeout = std::chrono::seconds{60};
```
@michael-redpanda 
i didn't wire this through the cluster and link... so it is not currently configurable.
Do we want it exposed?

michael-redpanda replied:
There is a `metadata_max_age_ms` option, I wonder if you could just double that and use it?

IoannisRP replied:
yeah, having it as a function of that is not bad.

Doubling sounds too little, though. It means that if 2 calls fail, we remove it. 3 sounds like a better redundancy...

IoannisRP committed:
kafka/client: make topic timeout duration a multiple of the max metadata age in cluster

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/e19867a7c8d734508fbbdae8d8ca2e58501a55b6..8781715b16401c56aefda7526030f6d496d7495c):
- wire topic timeout through max metadata age in cluster
